scene/animation/animation_tree.cpp

```diff
@@ -2203,6 +2208,80 @@ real_t AnimationTree::get_connection_activity(const StringName &p_path, int p_co
 	return (*activity)[p_connection].activity;
 }
 
+void AnimationTree::set_record_buffer_max_size(uint32_t p_max_size) {
+	record_buffer_max_size = p_max_size;
+
+	record_buffer.adjust_capacity(record_buffer_max_size);
+}
+
+void AnimationTree::resize_record_buffer(int p_size) {
+	record_buffer.resize(p_size);
+}
+
+void AnimationTree::discard_latest_records(uint32_t p_count) {
+	ERR_FAIL_COND(p_count > static_cast<uint32_t>(record_buffer.size()));
+
+	for (uint32_t i = 0; i < p_count; ++i) {
+		record_buffer.pop_front();
+	}
+}
+
+void AnimationTree::set_recording_rate(real_t p_rate) {
+	recording_rate = p_rate;
+	real_t min_rate = 1.0 / Engine::get_singleton()->get_frames_per_second();
+	if (recording_rate > min_rate) {
+		recording_rate = min_rate;
+	}
+}
+
+void AnimationTree::save_record(bool p_save_invalid_reasons) {
+	Record new_record;
+	new_record.state_animation_states.reserve(state.animation_states.size());
+	for (const AnimationNode::AnimationState &E : state.animation_states) {
+		new_record.state_animation_states.push_back(E);
+	}
+	new_record.state_last_pass = state.last_pass;
+	new_record.state_valid = state.valid;
+	new_record.process_pass = process_pass;
+	new_record.root_motion_position = root_motion_position;
+	new_record.root_motion_rotation = root_motion_rotation;
+	new_record.root_motion_scale = root_motion_scale;
+	new_record.root_motion_position_accumulator = root_motion_position_accumulator;
+	new_record.root_motion_rotation_accumulator = root_motion_rotation_accumulator;
+	new_record.root_motion_scale_accumulator = root_motion_scale_accumulator;
+	new_record.property_map = property_map;
+	if (p_save_invalid_reasons) {
+		new_record.state_invalid_reasons = state.invalid_reasons;
+	}
+
+	record_buffer.push_front(new_record);
```

Comment:
`push_front()` is high cost, so `push_back()` should be used to ensure that the end of the list is the most recent record.

Reply:
`record_buffer` using a new template `CircularDeque`, push element at edge is not high cost.